src/Microsoft.AspNetCore.Mvc.Formatters.Xml/XmlSerializerInputFormatter.cs

```diff
@@ -105,24 +111,42 @@ public int MaxDepth
                 request.Body.Seek(0L, SeekOrigin.Begin);
             }
 
-            using (var xmlReader = CreateXmlReader(new NonDisposableStream(request.Body), encoding))
+            try
             {
-                var type = GetSerializableType(context.ModelType);
+                using (var xmlReader = CreateXmlReader(new NonDisposableStream(request.Body), encoding))
+                {
+                    var type = GetSerializableType(context.ModelType);
 
-                var serializer = GetCachedSerializer(type);
+                    var serializer = GetCachedSerializer(type);
 
-                var deserializedObject = serializer.Deserialize(xmlReader);
+                    var deserializedObject = serializer.Deserialize(xmlReader);
 
-                // Unwrap only if the original type was wrapped.
-                if (type != context.ModelType)
+                    // Unwrap only if the original type was wrapped.
+                    if (type != context.ModelType)
+                    {
+                        if (deserializedObject is IUnwrappable unwrappable)
+                        {
+                            deserializedObject = unwrappable.Unwrap(declaredType: context.ModelType);
+                        }
+                    }
+
+                    return InputFormatterResult.Success(deserializedObject);
+                }
+            }
+            catch (Exception exception)
```

Comment:
`catch (InvalidOperationException exception) when (exception.InnerException is FormatException || exception.InnerException is XmlException)`